keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -232,10 +233,21 @@ However, the APIs have major differences:
 
 - Scalability and latency of DNS programming.
 
-  DNS programming requires the update of EndpointSlices and writing DNS records.
+  DNS programming requires the update of EndpointSlices by the endpoint
+  controller and updating DNS records by the DNS provider.
   This might not scale well for short-lived Jobs with high number of parallelism.
-  Moreoever, Pods need to be prepared to retry lookups in the case were the
-  records didn't have time to update.
+  Moreoever, Pods need to be prepared to:
+  - Retry lookups in the case were the records didn't have time to update.
+  - Handle more than one IP for the CNAME. This might happen temporarily when:
+    - the job controller creates more than one pod per index or
+    - the job controller creates a replacement of a failed Pod before the DNS
+      provider clears the record for the failed pod. This will be uncommon
+      as the endpoint controller should see the failed Pod before it sees the
+      replacement Pod.
+    <UNRESOLVED>
+    The recommendation for applications is to request a new DNS resolution until
+    the DNS server returns one IP.
```

Comment:
While we should point to the fact that a hostname may resolve to more than one IP, I am not sure about this recommendation because I don't think applications are built that way, selecting the IP is typically handled in the linux network stack. 

@johnbelamaric wdyt?

Reply:
The OS network stack returns a set of addresses https://golang.org/pkg/net/#LookupHost

It is the network libraries which decide what to do. In the case of the golang `net` package, it tries all of them until it finds one that doesn't fail https://github.com/golang/go/blob/109d7580a5c507b1e4f460445a5c4cd7313e4aa6/src/net/dial.go#L524

I think this is enough to handle the case "two IPs returned, one for a failed pod and one for new pod".
For the other case (more than one pod created per index), I don't have a proper alternative. But, recall that the job controller will try to remove the pod that was created/started later first. So the net package's algorithm seems like a reasonable solution for this too.

In conclusion, the strategy of using the first IP that doesn't fail seems reasonable to me. And if someone needs stronger guarantees, they can write their own Dialer that fails if it sees more than one IP. WDYT?

Reply:
cc @kow3ns @soltysh

Reply:
Something I didn't touch on is DNS caches. I don't see any indication of the `net` packaging caching DNS resolutions. So as long as user don't setup caches external to the application, they should be fine. And we can add as a recommendation to not set DNS caches for indexed jobs.

Do you have any other recommendations on this regard? I imagine users of Services face similar problems.

Reply:
That sounds reasonable to me; I think the key here is proper documentation to set expectations.

Reply:
> Do you have any other recommendations on this regard? I imagine users of Services face similar problems.

Same thing, proper documentation and stressing the point that pods are ephemeral and may change IPs when getting recreated, this could happen for StatefulSets as well.

Reply:
I read Aldo's golang comment as just an example for the sake of this discussion, the general pattern in other languages is that there are two functions, one to resolve the hostname and obtain the addresses, and one that accepts one of those addresses to establish a connection (this is the same in C/C++ socket programming). We certainly shouldn't be relying on the semantics of any specific language. But yeah, as I mentioned above, I don't think we should make the recommendation mentioned in the text and simply stress that pods are ephemeral and may change IPs when getting recreated and so applications needs to be tolerant to that.

Reply:
Exactly - that was what I was trying to push us towards. Discuss what you mentioned above but don't provide a concrete recommendation.

Reply:
The SIG Apps leads were asking for a specific recommendation.

I've changed the text to describe what happens when there are 2 pods per index. I added the explicit note that DNS caches shouldn't be used. I hope this is good enough.

Also note that I'm not sure of the circumstances that would lead to more than one pod per index. In a high level, this happens when the job controller misses a Pod creation event. But:
- we keep an in-memory track of creations issued (which we call expectations) and we don't create or delete pods until those creations are observed.
- let's say the controller restarts due to a crash or reboot. The new controller waits for an informer cache sync before processing jobs.

Perhaps the only scenario would be were an API request connection drops, but the apiserver successfully created an object. In this case, the controller would consider this as a failure and wouldn't add this creation to the expectations. Are there other scenarios?